omniscidb/QueryEngine/ColumnFetcher.cpp

```diff
@@ -262,44 +267,96 @@ const int8_t* ColumnFetcher::getAllTableColumnFragments(
-    } else {
-      table_column = column_it->second.get();
+
+      return ColumnFetcher::transferColumnIfNeeded(
+          table_column, 0, memory_level, device_id, device_allocator);
+    }
+
+    const auto type_width = col_info->type->size();
+    auto write_ptr =
+        executor_->row_set_mem_owner_->allocate(type_width * total_row_count);
+    std::vector<std::pair<int8_t*, size_t>> write_ptrs;
+    std::vector<size_t> valid_fragments;
+    for (size_t frag_id = 0; frag_id < frag_count; ++frag_id) {
+      const auto& fragment = (*fragments)[frag_id];
+      if (fragment.isEmptyPhysicalFragment()) {
+        continue;
+      }
+      CHECK_EQ(type_width, fragment.getChunkMetadataMap().at(col_id)->type()->size());
+      write_ptrs.push_back({write_ptr, fragment.getNumTuples() * type_width});
+      write_ptr += fragment.getNumTuples() * type_width;
+      valid_fragments.push_back(frag_id);
     }
+
+    CHECK(!write_ptrs.empty());
+    size_t valid_frag_count = valid_fragments.size();
+    tbb::parallel_for(
+        tbb::blocked_range<size_t>(0, valid_frag_count),
+        [&](const tbb::blocked_range<size_t>& frag_ids) {
+          for (size_t v_frag_id = frag_ids.begin(); v_frag_id < frag_ids.end();
+               ++v_frag_id) {
+            std::list<std::shared_ptr<Chunk_NS::Chunk>> chunk_holder;
+            std::list<ChunkIter> chunk_iter_holder;
+            size_t frag_id = valid_fragments[v_frag_id];
+            const auto& fragment = (*fragments)[frag_id];
+            auto chunk_meta_it = fragment.getChunkMetadataMap().find(col_id);
+            CHECK(chunk_meta_it != fragment.getChunkMetadataMap().end());
+            std::shared_ptr<Chunk_NS::Chunk> chunk;
+            {
+              ChunkKey chunk_key{
+                  db_id, fragment.physicalTableId, col_id, fragment.fragmentId};
+              chunk = data_provider_->getChunk(col_info,
+                                               chunk_key,
+                                               Data_Namespace::CPU_LEVEL,
+                                               0,
+                                               chunk_meta_it->second->numBytes(),
+                                               chunk_meta_it->second->numElements());
+              std::lock_guard<std::mutex> chunk_list_lock(chunk_list_mutex_);
+              chunk_holder.push_back(chunk);
+            }
+            auto ab = chunk->getBuffer();
+            CHECK(ab->getMemoryPtr());
+            int8_t* col_buffer =
+                ab->getMemoryPtr();  // @TODO(alex) change to use ChunkIter
+            memcpy(write_ptrs[frag_id].first, col_buffer, write_ptrs[frag_id].second);
+          }
+        });
+
+    std::vector<int8_t*> raw_write_ptrs;
+    raw_write_ptrs.reserve(frag_count);
+    for (size_t i = 0; i < frag_count; i++) {
+      raw_write_ptrs.emplace_back(write_ptrs[i].first);
+    }
+
+    std::unique_ptr<ColumnarResults> merged_results(new ColumnarResults(
```

Comment:
The vector `ColumnarResults::ColumnarResults` gets is not a pointer per fragment, it is a pointer per column (`ColumnarResults` can store multiple columns, but not multiple fragments). So actually you are supposed to pass a vector with a single pointer here. Your version works because the first vector element has a correct pointer and other elements are simply not used. But this code is still confusing, so let's fix it.

Reply:
I don't understand. Do you mean that we are using only data from first fragment? If so why we are fetching them?

Reply:
Ah, sorry, now I get it.